### ISSUE TITLE
RORem Discriminator

### ISSUE THREAD

**Issue by amindehnavi** [leeruibin/RORem#5]
First of all, thank you for your interesting work and for making it publicly available. After cloning the repo to test the results, I faced some strange results.

I have a test bench containing about 90 images for object removal. I took inferences from your model and some other SOTA methods on my test data, and then I wanted to evaluate the results using the RORem Discriminator model.  After checking the results, I found out that the discriminator's performance is very good when I use the RORem-mixed model to remove objects, but it is less accurate when I check the results, obtained from other SOTA models. 

For instance, consider the following examples:

Input Image and Mask:

![Image](https://github.com/user-attachments/assets/43846a6e-8374-4a22-be56-0c62ccfce4ab)

![Image](https://github.com/user-attachments/assets/ed8fd23f-1015-4e2b-87dc-3db997583228)

The result of taking inference from the RORem model and the rorem discriminator:

![Image](https://github.com/user-attachments/assets/26c26ed8-5106-46ad-8d1b-f21302247158)

`predicted_score = 0.82861328125`

The result of taking inference from another method and the rorem discriminator:

![Image](https://github.com/user-attachments/assets/90955b56-32c1-44fe-a9df-7776735bd547)

`predicted_score = 0.11773681640625`

Do you have any idea why this happened? 

As you mentioned in the section "4.3. Training and Evaluation of Discriminator" in your paper, I expected the RORem discriminator would not be biased on the output distribution of the RORem model. Does this relate to being biased on the RORem output distribution?

**Comment by leeruibin** [comment]
Thank you for your interest. May I ask which method you are currently using?

In our experiments, we primarily evaluate the performance of the discriminator on methods presented in our paper. Although we have expanded the training dataset to enhance the discriminator's robustness, its performance may still degrade when encountering new removal patterns.

Additionally, since our discriminator is trained on images with a resolution of 512×512, using images of other sizes may also lead to a decline in model performance.

**Comment by amindehnavi** [comment]
> Thank you for your interest. May I ask which method you are currently using?
> 
> In our experiments, we primarily evaluate the performance of the discriminator on methods presented in our paper. Although we have expanded the training dataset to enhance the discriminator's robustness, its performance may still degrade when encountering new removal patterns.
> 
> Additionally, since our discriminator is trained on images with a resolution of 512×512, using images of other sizes may also lead to a decline in model performance.

Thank you for your reply. I used my own model to test the result. my model is actually the SDXL Inpainting model which is trained on the RORD and Mulan dataset, similar your round 1 in the paper. 

About the resolution of 512x512, as far as I found out, in the `inference_RORem_discriminator.py` you resize the inputs before passing them to the discriminator:

```
source_image = load_image(args.image_path).resize([512, 512])
mask_image = load_image(args.mask_path).resize([512, 512])
GT_image = load_image(args.edited_path).resize([512, 512])
```

**Comment by leeruibin** [comment]
Did you perform center cropping before feeding the images into the discriminator? All my test data are square-shaped, so resizing won't cause distortion. But the images you provided are rectangular - direct resizing might lead to unwanted stretching or aspect ratio issues.

**Comment by amindehnavi** [comment]
As you said, when I perform center cropping, the discriminator results get better. However, there are still some images in the predicted class by discriminator that vary between the results obtained from the Rorem model and ours. here, the point is that the difference between the predicted scores is considerable. For example, see this sample:

### Input Image and Mask

![Image](https://github.com/user-attachments/assets/4df8178f-4f64-4f62-a96b-679ce24c106d)

![Image](https://github.com/user-attachments/assets/b41dafd1-48e7-4680-bd21-c60aff407ff1)

### RORem
`Prediction_score = 0.84619140625`
![Image](https://github.com/user-attachments/assets/e6c8fdda-3094-4360-b1bf-7a86f6bdf0b6)

### Ours 
`Prediction_score = 0.11053466796875`
![Image](https://github.com/user-attachments/assets/85dbfed4-3506-48e2-abbf-71cc20b799f0)

**Comment by leeruibin** [comment]
Whether the target is successfully removed at certain times is a somewhat subjective issue. In the case you presented, we tend to consider it a failure because irregular license plates and a distorted car bumper appear at the center of the vehicle. When we had the annotators label it, they might have regarded this as an artifact in the background and thus given a negative rating. As a result, the discriminator trained on such data would also evaluate it negatively. You can fine-tune the model based on your own judgment data, but for now, we lean toward classifying this image as a negative sample.

![Image](https://github.com/user-attachments/assets/76124226-ad69-4c76-a4d2-8510b1f86f6e)

**Comment by amindehnavi** [comment]
Thank you for your quick response. 

I agree with you that the discriminator may be sensitive to these small artifacts and distortions. In fact, I wonder why it should be so sensitive, lowering the prediction score from 0.84 to 0.1. This idea comes from the fact that I checked the RORem dataset that you published, and I saw similar artifacts in that dataset, too. As your discriminator is trained on these data, I expect it to be more robust against such distortions and artifacts. Consider this example from your dataset:


### Input Image and Mask

![Image](https://github.com/user-attachments/assets/a8cae63e-7ebe-4f5c-9ead-a06cd348b75a)

![Image](https://github.com/user-attachments/assets/29ef077d-6dc4-4e4a-a8fa-0a856d021608)

### Inpainted Image

![Image](https://github.com/user-attachments/assets/0cb4ab16-5553-4ce7-9ae1-f0c362960a4c)

`Predicted Score = 0.9033203125`


As you mentioned, I think that I need to fine-tune it on my model's output to get the best results.

**Comment by leeruibin** [comment]
You could expand the human-annotated dataset to further fine-tune the discriminator for better performance.

Meanwhile, note that we do not train the discriminator on the entire RORem dataset, as manually annotating all 200K samples would be prohibitively time-consuming and costly. Instead, our discriminator is trained on 30K human-annotated samples, supplemented with augmented data as mentioned in our paper.

**Comment by amindehnavi** [comment]
I am grateful for your detailed explanations and thoughtful responses. They are sincerely appreciated.
